pkg/kubelet/mirror_manager.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+// Kubelet discover pod updates from 4 sources: file, http, etcd, and
+// apiserver. Pods from non-apiserver sources are called static pods, and API
+// server is not aware of the existence of static pods. In order to monitor the
+// status of such pods, kubelet create a mirror pod for each static pod via the
+// API server.
+//
+// A mirror pod has the same pod full name (name and namespace) as its static
+// counterpart (albeit different metadata such as UID, etc). By leveraging the
+// fact that kubelet reports the pod status using the pod full name, the status
+// of the mirror pod always reflects the acutal status of the static pod.
+// When a static pod gets deleted, the associated orphaned mirror pods will
+// also be removed.
+//
+// This file includes functions to manage the mirror pods.
+
+type mirrorManager interface {
+	CreateMirrorPod(*api.BoundPod, string) error
+	DeleteMirrorPod(string) error
+}
+
+type basicMirrorManager struct {
+	// mirror pods are stored in the kubelet directly because they need to be
+	// in sync with the internal pods.
+	apiserverClient client.Interface
+}
+
+func newBasicMirrorManager(apiserverClient client.Interface) *basicMirrorManager {
+	return &basicMirrorManager{apiserverClient: apiserverClient}
+}
+
+// Creates a mirror pod.
+func (self *basicMirrorManager) CreateMirrorPod(pod *api.BoundPod, hostname string) error {
+	if self.apiserverClient == nil {
+		return nil
+	}
+	podFullName := GetPodFullName(pod)
+	// Create a Pod from BoundPod; this will become redundant once we
+	// deprecate BoundPod completely.
+	// Note that some fields (e.g. creationTimestamp, UID, etc) will be
+	// overwritten by the APIserver.
+	podToCreate := api.Pod{}
+	if err := api.Scheme.Convert(pod, &podToCreate); err != nil {
+		glog.Errorf("Unable to interpret static bound pod as a pod: %v: %+v", err, pod)
+	}
+	// Indicate that the pod should be scheduled to the current node.
+	podToCreate.Spec.Host = hostname
+	podToCreate.Annotations[ConfigMirrorAnnotationKey] = MirrorType
+
+	glog.V(4).Infof("Creating a mirror pod %q", podFullName)
+	_, err := self.apiserverClient.Pods(NamespaceDefault).Create(&podToCreate)
+	if err != nil {
+		glog.Errorf("Failed creating a mirror pod %q: %#v", podFullName, err)
+		return err
+	}
+	return nil
+}
+
+// Deletes a mirror pod.
+func (self *basicMirrorManager) DeleteMirrorPod(podFullName string) error {
+	if self.apiserverClient == nil {
+		return nil
+	}
+	name, namespace, err := parsePodFullName(podFullName)
+	if err != nil {
+		return err
+	}
+	glog.V(4).Infof("Deleting a mirror pod %q", podFullName)
+	if err := self.apiserverClient.Pods(namespace).Delete(name); err != nil {
+		glog.Errorf("Failed deleting a mirror pod %q: %v", podFullName, err)
+	}
+	return nil
+}
+
+// Delete all orphaned mirror pods.
+func (kl *Kubelet) deleteOrphanedMirrorPods(pods []api.BoundPod, mirrorPods util.StringSet) {
```

Comment:
We can probably make this a member of mirrorManager and have Kubelet call it since it doesn't use any special Kubelet state.

A pet peeve of mine is also to keep all of a struct's methods in one file, but that's just me :)